File: app/src/main/java/com/veganbeauty/admin/MainActivity.java
package com.veganbeauty.admin;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import com.google.firebase.firestore.FieldValue;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.ListenerRegistration;
import com.google.firebase.firestore.SetOptions;
import com.google.firebase.messaging.FirebaseMessaging;
import com.veganbeauty.admin.core.base.RootieAdminActivity;
import com.veganbeauty.admin.data.local.SessionManager;
import com.veganbeauty.admin.data.remote.ChatMessage;
import com.veganbeauty.admin.data.remote.FirebaseService;
import com.veganbeauty.admin.databinding.ActivityMainBinding;
import com.veganbeauty.admin.features.auth.LoginActivity;
import com.veganbeauty.admin.features.home.BottomNavHelper;
import com.veganbeauty.admin.features.home.HomeFragment;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class MainActivity extends RootieAdminActivity {

    public interface ChatMessagesListener {
        void onChatMessagesUpdated(List<ChatMessage> messages);
    }

    private ActivityMainBinding binding;
    private ListenerRegistration chatListenerRegistration = null;
    private final FirebaseService firebaseService = new FirebaseService();
    private final List<ChatMessagesListener> chatMessagesListeners = new ArrayList<>();
    private List<ChatMessage> cachedChatMessages = new ArrayList<>();
    private int lastUnreadCount = 0;
    private int currentTabId = R.id.nav_home;

    public int getLastUnreadCount() {
        return lastUnreadCount;
    }

    public int getCurrentTabId() {
        return currentTabId;
    }

    public void setCurrentTabId(int currentTabId) {
        this.currentTabId = currentTabId;
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        SessionManager sessionManager = new SessionManager(this);
        if (!sessionManager.isLoggedIn()) {
            Intent intent = new Intent(this, LoginActivity.class);
            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
            startActivity(intent);
            finish();
            super.onCreate(savedInstanceState);
            return;
        }

        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        super.onCreate(savedInstanceState);

        if (savedInstanceState == null) {
            loadFragment(new HomeFragment());
        }

        getSupportFragmentManager().addOnBackStackChangedListener(() -> {
            if (getSupportFragmentManager().getBackStackEntryCount() == 0) {
                ensureBottomNavVisible();
            }
        });

        startChatUnreadListener();
        requestNotificationPermission();
        logFCMToken();
    }

    private void requestNotificationPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.POST_NOTIFICATIONS) !=
                PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(
                    this,
                    new String[]{Manifest.permission.POST_NOTIFICATIONS},
                    101
                );
            }
        }
    }

    private void logFCMToken() {
        try {
            FirebaseMessaging.getInstance().getToken().addOnCompleteListener(task -> {
                if (!task.isSuccessful()) {
                    Log.w("FCM", "Fetching FCM registration token failed", task.getException());
                    return;
                }
                String token = task.getResult();
                Log.d("FCM", "==================================================");
                Log.d("FCM", "Current FCM Token: " + token);
                Log.d("FCM", "==================================================");
                saveTokenToFirestore(token);
            });
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void saveTokenToFirestore(String token) {
        try {
            FirebaseFirestore db = FirebaseFirestore.getInstance();
            Map<String, Object> data = new HashMap<>();
            data.put("token", token);
            data.put("updated_at", FieldValue.serverTimestamp());

            db.collection("admin_metadata").document("fcm")
                .set(data, SetOptions.merge())
                .addOnSuccessListener(aVoid -> Log.d("FCM", "FCM Token saved to Firestore successfully"))
                .addOnFailureListener(e -> Log.e("FCM", "FCM Token failed to save to Firestore", e));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    protected void setupUI() {
        SessionManager sessionManager = new SessionManager(this);
        String role = sessionManager.getRole();
        if (role == null) {
            role = "admin";
        }

        View bottomNav = findViewById(R.id.bottom_nav);
        if (bottomNav != null) {
            View navProduct = bottomNav.findViewById(R.id.nav_product);
            View navCustomer = bottomNav.findViewById(R.id.nav_customer);
            if ("staff".equals(role) || "nhân viên".equals(role)) {
                if (navProduct != null) navProduct.setVisibility(View.GONE);
                if (navCustomer != null) navCustomer.setVisibility(View.GONE);
            } else {
                if (navProduct != null) navProduct.setVisibility(View.VISIBLE);
                if (navCustomer != null) navCustomer.setVisibility(View.VISIBLE);
            }

            BottomNavHelper.setup(
                this,
                bottomNav,
                currentTabId,
                tabId -> {
                    currentTabId = tabId;
                    BottomNavHelper.navigate(this, tabId);
                }
            );
            BottomNavHelper.highlightTab(bottomNav, currentTabId);
        }
    }

    public void loadFragment(Fragment fragment) {
        ensureBottomNavVisible();
        getSupportFragmentManager().beginTransaction()
            .replace(binding.mainContainer.getId(), fragment)
            .commit();
        bringBottomNavToFront();
    }

    public void loadFragmentHidingNav(Fragment fragment) {
        hideBottomNav();
        getSupportFragmentManager().beginTransaction()
            .replace(binding.mainContainer.getId(), fragment)
            .addToBackStack(null)
            .commit();
    }

    public void ensureBottomNavVisible() {
        View bottomNav = findViewById(R.id.bottom_nav);
        if (bottomNav != null) {
            bottomNav.setVisibility(View.VISIBLE);
            bringBottomNavToFront();
        }
    }

    public void hideBottomNav() {
        View bottomNav = findViewById(R.id.bottom_nav);
        if (bottomNav != null) {
            bottomNav.setVisibility(View.GONE);
        }
    }

    private void bringBottomNavToFront() {
        View bottomNav = findViewById(R.id.bottom_nav);
        if (bottomNav != null) {
            bottomNav.bringToFront();
            bottomNav.requestLayout();
        }
    }

    public void addChatMessagesListener(ChatMessagesListener listener) {
        if (listener == null || chatMessagesListeners.contains(listener)) {
            return;
        }
        chatMessagesListeners.add(listener);
        if (!cachedChatMessages.isEmpty()) {
            listener.onChatMessagesUpdated(new ArrayList<>(cachedChatMessages));
        }
    }

    public void removeChatMessagesListener(ChatMessagesListener listener) {
        chatMessagesListeners.remove(listener);
    }

    public List<ChatMessage> getCachedChatMessages() {
        return new ArrayList<>(cachedChatMessages);
    }

    private void notifyChatMessagesListeners(List<ChatMessage> messages) {
        List<ChatMessage> copy = new ArrayList<>(messages);
        runOnUiThread(() -> {
            for (ChatMessagesListener listener : new ArrayList<>(chatMessagesListeners)) {
                listener.onChatMessagesUpdated(copy);
            }
        });
    }

    private void startChatUnreadListener() {
        SessionManager sessionManager = new SessionManager(this);
        if (!sessionManager.isLoggedIn()) return;

        if (chatListenerRegistration != null) {
            chatListenerRegistration.remove();
        }

        try {
            FirebaseFirestore firestore = FirebaseFirestore.getInstance();
            chatListenerRegistration = firestore.collection("community_message")
                .addSnapshotListener((snapshot, e) -> {
                    if (e != null) {
                        e.printStackTrace();
                        return;
                    }

                    if (snapshot != null) {
                        List<ChatMessage> messages = firebaseService.parseChatMessagesFromSnapshot(snapshot);
                        cachedChatMessages = messages;
                        int unreadConvCount = firebaseService.countUnreadConversations(snapshot);
                        lastUnreadCount = unreadConvCount;
                        updateGlobalMessageBadges(unreadConvCount);
                        notifyChatMessagesListeners(messages);
                    }
                });
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }

    public void updateGlobalMessageBadges(int count) {
        TextView homeBadge = findViewById(R.id.home_header_message_badge);
        TextView otherBadge = findViewById(R.id.message_badge);

        updateBadgeTextView(homeBadge, count);
        updateBadgeTextView(otherBadge, count);
    }

    private void updateBadgeTextView(TextView badgeView, int count) {
        if (badgeView == null) return;
        if (count > 0) {
            badgeView.setText(String.valueOf(count));
            badgeView.setVisibility(View.VISIBLE);
        } else {
            badgeView.setVisibility(View.GONE);
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (chatListenerRegistration != null) {
            chatListenerRegistration.remove();
        }
    }
}

File: app/src/main/java/com/veganbeauty/admin/MainActivity.kt
package com.veganbeauty.admin

import android.content.Intent
import android.content.Context
import android.os.Bundle
import android.view.View
import androidx.fragment.app.Fragment
import com.veganbeauty.admin.core.base.RootieAdminActivity
import com.veganbeauty.admin.data.local.SessionManager
import com.veganbeauty.admin.databinding.ActivityMainBinding
import com.veganbeauty.admin.features.auth.LoginActivity
import com.veganbeauty.admin.features.home.BottomNavHelper
import com.veganbeauty.admin.features.home.HomeFragment
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.ViewModelProvider
import com.veganbeauty.admin.features.booking.BookingViewModel
import com.veganbeauty.admin.data.local.entities.BookingEntity
import com.veganbeauty.admin.data.local.RootieAdminDatabase
import androidx.appcompat.app.AlertDialog
import android.media.RingtoneManager
import android.os.Vibrator
import android.os.VibrationEffect
import android.os.Build

class MainActivity : RootieAdminActivity() {

    private lateinit var binding: ActivityMainBinding
    private lateinit var bookingViewModel: BookingViewModel
    
    val pendingBookingsCount = MutableLiveData<Int>(0)
    val pendingBookingCount = pendingBookingsCount // Alias for compatibility
    val pendingOrdersCount = MutableLiveData<Int>(0)
    val unreadChatsCount = MutableLiveData<Int>(0)
    val totalNotificationCount = MediatorLiveData<Int>().apply {
        addSource(pendingBookingsCount) { value = (pendingBookingsCount.value ?: 0) + (pendingOrdersCount.value ?: 0) + (unreadChatsCount.value ?: 0) }
        addSource(pendingOrdersCount) { value = (pendingBookingsCount.value ?: 0) + (pendingOrdersCount.value ?: 0) + (unreadChatsCount.value ?: 0) }
        addSource(unreadChatsCount) { value = (pendingBookingsCount.value ?: 0) + (pendingOrdersCount.value ?: 0) + (unreadChatsCount.value ?: 0) }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        val sessionManager = SessionManager(this)
        if (!sessionManager.isLoggedIn()) {
            val intent = Intent(this, LoginActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
            startActivity(intent)
            finish()
            super.onCreate(savedInstanceState)
            return
        }

        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        super.onCreate(savedInstanceState)
        
        bookingViewModel = ViewModelProvider(this)[BookingViewModel::class.java]
        
        // Start real-time Firestore sync
        bookingViewModel.startRealtimeSync { newBooking ->
            showNewBookingDialog(newBooking)
        }
        
        // Observe bookings database changes to update badge count
        bookingViewModel.allBookings.observe(this) { bookings ->
            updatePendingBookingCount(bookings)
        }
        
        val db = RootieAdminDatabase.getDatabase(this)
        
        // Observe orders database changes to update pending orders count
        db.orderDao().getAllLiveData().observe(this) { orders ->
            val count = orders.filter { it.status == "Chờ xử lý" }.size
            pendingOrdersCount.postValue(count)
        }
        
        // Observe chat messages to update unread chats count
        db.chatMessageDao().getAllMessages().observe(this) { messages ->
            val count = messages.filter { !it.isRead && it.sender == "customer" }.map { it.threadId }.distinct().size
            unreadChatsCount.postValue(count)
        }

        if (savedInstanceState == null) {
            loadFragment(HomeFragment())
        }
    }
    var currentTabId: Int = R.id.nav_home

    override fun setupUI() {
        val sessionManager = SessionManager(this)
        val role = sessionManager.getRole() ?: "admin"

        val bottomNav = findViewById<View>(R.id.bottom_nav)
        if (bottomNav != null) {
            if (role == "staff" || role == "nhân viên") {
                bottomNav.findViewById<View>(R.id.nav_product)?.visibility = View.GONE
                bottomNav.findViewById<View>(R.id.nav_customer)?.visibility = View.GONE
            } else {
                bottomNav.findViewById<View>(R.id.nav_product)?.visibility = View.VISIBLE
                bottomNav.findViewById<View>(R.id.nav_customer)?.visibility = View.VISIBLE
            }

            BottomNavHelper.setup(
                activity = this,
                root = bottomNav
            ) { tabId ->
                // Luôn cho phép navigate về Home, dù đang ở Home hay không
                BottomNavHelper.navigate(this, tabId)
            }
            BottomNavHelper.highlightTab(bottomNav, currentTabId)
        }
    }

    /**
     * Được gọi bởi các fragment con khi tự điều hướng (không qua bottom nav)
     * để đồng bộ trạng thái highlight của bottom nav.
     */
    fun syncTab(tabId: Int) {
        currentTabId = tabId
        val bottomNav = findViewById<View>(R.id.bottom_nav) ?: return
        BottomNavHelper.highlightTab(bottomNav, tabId)
    }

    fun loadFragment(fragment: Fragment) {
        supportFragmentManager.beginTransaction()
            .replace(binding.mainContainer.id, fragment)
            .commit()
    }

    private fun updatePendingBookingCount(bookings: List<BookingEntity>) {
        val sessionManager = SessionManager(this)
        val role = sessionManager.getRole() ?: "admin"
        val storeID = sessionManager.getStoreID() ?: ""
        val isStaff = role.equals("staff", ignoreCase = true) || role.equals("nhân viên", ignoreCase = true)
        
        val count = bookings.filter { booking ->
            val isPending = booking.status.equals("Chờ xác nhận", ignoreCase = true) || 
                            booking.status.equals("pending", ignoreCase = true)
            
            val isTargetBranch = if (isStaff && storeID.isNotEmpty()) {
                booking.storeID == storeID || (booking.storeID.isEmpty() && (
                    (storeID == "CH001" && booking.storeName.contains("Cơ sở 1", ignoreCase = true)) ||
                    (storeID == "CH005" && booking.storeName.contains("Cơ sở 5", ignoreCase = true))
                ))
            } else {
                // Admin matches both CH001 and CH005
                booking.storeID == "CH001" || booking.storeID == "CH005" ||
                booking.storeName.contains("Cơ sở 1", ignoreCase = true) || 
                booking.storeName.contains("Cơ sở 5", ignoreCase = true)
            }
            isPending && isTargetBranch
        }.size
        
        pendingBookingCount.postValue(count)
    }

    private fun showNewBookingDialog(booking: BookingEntity) {
        // Play notification sound
        try {
            val notificationUri = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION)
            val ringtone = RingtoneManager.getRingtone(applicationContext, notificationUri)
            ringtone?.play()
        } catch (e: Exception) {
            e.printStackTrace()
        }

        // Vibrate device
        try {
            val vibrator = getSystemService(Context.VIBRATOR_SERVICE) as? Vibrator
            if (vibrator != null && vibrator.hasVibrator()) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    vibrator.vibrate(VibrationEffect.createOneShot(500, VibrationEffect.DEFAULT_AMPLITUDE))
                } else {
                    @Suppress("DEPRECATION")
                    vibrator.vibrate(500)
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun onDestroy() {
        if (::bookingViewModel.isInitialized) {
            bookingViewModel.stopRealtimeSync()
        }
        super.onDestroy()
    }
}
